From 97b7aa13810e9796770351c36cab75492f02058e Mon Sep 17 00:00:00 2001
From: Vladimir Panteleev <git@cy.md>
Date: Sat, 20 Dec 2025 12:31:48 +0000
Subject: [PATCH] sys.file: Use checked cast to off_t in allocate()

This fixes compilation on 32-bit platforms where off_t may be 32-bit,
while still detecting overflow at runtime if the length exceeds the
off_t range.
---
 sys/file.d | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/sys/file.d b/sys/file.d
index da183c0f..de5c6e5a 100644
--- a/sys/file.d
+++ b/sys/file.d
@@ -2470,9 +2470,11 @@ void allocate(File f, ulong length)
 	version (linux)
 	{
 		import core.sys.linux.fcntl : fallocate;
+		import core.sys.posix.sys.types : off_t;
+		import std.conv : to;
 		f.flush();
 		// TODO: add overloads for other parameters
-		auto res = fallocate(f.fileno, 0, 0, length);
+		auto res = fallocate(f.fileno, 0, 0, length.to!off_t);
 		errnoEnforce(res == 0, "fallocate");
 	}
 	else
-- 
2.49.0

